src/libNode/Node.cpp

```diff
@@ -2408,6 +2427,199 @@ bool Node::ProcessDoRejoin(const bytes& message, unsigned int offset,
+    lock_guard<mutex> g(m_mutexShardMember);
+    for (auto& it : *m_myShardMembers) {
+      peerInfo.push_back(it.second);
+    }
+  }
+
+  {
+    // Multicast to all DS committee
+    lock_guard<mutex> lock(m_mediator.m_mutexDSCommittee);
+    for (auto const& i : *m_mediator.m_DSCommittee) {
+      if (i.second.m_listenPortHost != 0) {
+        peerInfo.push_back(i.second);
+      }
+    }
+  }
+
+  P2PComm::GetInstance().SendMessage(peerInfo, updateshardguardidentitymessage);
+
+  return true;
+}
+
+bool Node::ProcessNewShardGuardNetworkInfo(const bytes& message,
+                                           unsigned int offset,
+                                           const Peer& from) {
+  LOG_MARKER();
+
+  if (!GUARD_MODE) {
+    LOG_GENERAL(
+        WARNING,
+        "Not in guard mode. Unable to update shard guard network info.");
+    return false;
+  }
+
+  uint64_t dsEpochNumber;
+  Peer shardGuardNewNetworkInfo;
+  uint64_t timestamp;
+  PubKey shardGuardPubkey;
+
+  if (!Messenger::GetNodeNewShardGuardNetworkInfo(
+          message, offset, dsEpochNumber, shardGuardNewNetworkInfo, timestamp,
+          shardGuardPubkey)) {
+    LOG_EPOCH(WARNING, m_mediator.m_currentEpochNum,
+              "Messenger::GetNodeNewShardGuardNetworkInfo failed.");
+    return false;
+  }
+
+  if (from.GetIpAddress() != shardGuardNewNetworkInfo.GetIpAddress()) {
+    LOG_CHECK_FAIL("IP Address",
+                   shardGuardNewNetworkInfo.GetPrintableIPAddress(),
+                   from.GetPrintableIPAddress());
+    return false;
+  }
+
+  if (m_mediator.m_selfKey.second == shardGuardPubkey) {
+    LOG_GENERAL(INFO,
+                "[update shard guard] Node to be updated is current node. No "
+                "update needed.");
+    return false;
+  }
+
+  uint64_t currentDSEpochNumber =
+      m_mediator.m_dsBlockChain.GetLastBlock().GetHeader().GetBlockNum() + 1;
+
+  if (dsEpochNumber != currentDSEpochNumber) {
+    LOG_GENERAL(
+        WARNING,
+        "Update of shard guard network info failure  - dsepoch in message: "
+            << dsEpochNumber
+            << " does not match current dsepoch: " << currentDSEpochNumber);
+    return false;
+  }
+
+  if (!Guard::GetInstance().IsNodeInShardGuardList(shardGuardPubkey)) {
+    LOG_GENERAL(WARNING,
+                "Update of shard guard requested by non-shard guard node, So "
+                "ignore the request");
+    return false;
+  }
+  // I am lookup node
+  if (LOOKUP_NODE_MODE) {
+    m_mediator.m_ds->UpdateShardNodeNetworkInfo(shardGuardNewNetworkInfo,
+                                                shardGuardPubkey);
+    if (!BlockStorage::GetBlockStorage().PutShardStructure(
+            m_mediator.m_ds->m_shards, 0)) {
+      LOG_GENERAL(WARNING, "BlockStorage::PutShardStructure failed");
+    }
+  }
+  // I am sharded node and requestor is also from my shard
+  else if (m_mediator.m_ds->m_mode == DirectoryService::IDLE) {
+    if (!IsShardNode(shardGuardPubkey)) {
```

Comment:
Here we make an initial search through `m_myShardMembers` and then in line 2575 we also go through `m_myShardMembers` to make the actual replacement. Maybe we can skip this initial search and just try to find and replace immediately? If it's not found during replacement, you can log the "does not match" message as well.